network/src/protocols/discovery/state.rs

```diff
@@ -35,15 +35,26 @@ pub struct SessionState {
 }
 
 impl SessionState {
-    pub(crate) fn new(context: ProtocolContextMutRef) -> SessionState {
+    pub(crate) fn new<M: AddressManager>(
+        context: ProtocolContextMutRef,
+        addr_manager: &M,
+    ) -> SessionState {
         let mut addr_known = AddrKnown::default();
         let remote_addr = if context.session.ty.is_outbound() {
             let port = context
                 .listens()
                 .iter()
-                .filter_map(|address| multiaddr_to_socketaddr(address))
-                .map(|addr| addr.port())
-                .next();
+                .map(|address| {
+                    // Verify self is a public node first
+                    // if not, try to make public network nodes broadcast hole punching information
+                    if addr_manager.is_valid_addr(address) {
+                        multiaddr_to_socketaddr(address).map(|socket_addr| socket_addr.port())
+                    } else {
+                        None
+                    }
+                })
+                .next()
+                .flatten();
```

Comment:
`.map().next().flatten()` is different from `.filter_map().next()`. The former always return the first address map result.

Reply:
Maybe I should use `flat_map().next()`, now next returns an `Option<Option<u16>>`, use `flatten` to flattens nested `Option`